2.x: Flowable.groupBy add overload with evicting map factory

davidmoten opened the pull request:
As per discussion in #5763, this PR adds an overload for `Flowable.groupBy` that specifies an `evictingMapFactory`. 

An example of usage taken from the new javadoc:

```java
Function<Consumer<Object>, Map<Integer, Object>> evictingMapFactory = 
   notify ->
       CacheBuilder
         .newBuilder() 
         .maximumSize(3)
         .removalListener(entry -> { 
              try {
                  // emit the value not the key!
                  notify.accept(entry.getValue());
              } catch (Exception e) {
                  throw new RuntimeException(e);
              }
            })
         .<Integer, Object> build()
         .asMap();
         
 // Emit 1000 items but ensure that the
 // internal map never has more than 3 items in it           
 Flowable
   .range(1, 1000)
   // note that number of keys is 10
   .groupBy(x -> x % 10, x-> x, true, 16, evictingMapFactory)
   .flatMap(g -> g)
   .forEach(System.out::println);
```

Note that I based this operator on the 1.x version which I think may have a bug that goes as far as the signature of that method. The eviction consumer should not be given a key but rather the value from the map. I'll pursue the 1.x issue after dealing with this.

davidmoten committed:
2.x: groupBy add overload with evicting map factory

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
 
+    /**
+     * Groups the items emitted by a {@code Publisher} according to a specified criterion, and emits these
+     * grouped items as {@link GroupedFlowable}s. The emitted {@code GroupedPublisher} allows only a single
```
`GroupedPublisher` -> `GroupedFlowable`

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * grouped items as {@link GroupedFlowable}s. The emitted {@code GroupedPublisher} allows only a single
+     * {@link Subscriber} during its lifetime and if this {@code Subscriber} cancels before the
+     * source terminates, the next emission by the source having the same key will trigger a new
+     * {@code GroupedPublisher} emission. The {@code evictingMapFactory} is used to create a map that will
```
`GroupedPublisher` -> `GroupedFlowable`

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * 
+     * <p>A use case for specifying an {@code evictingMapFactory} is where the source is infinite and fast and
+     * over time the number of keys grows enough to be a concern in terms of the memory footprint of the 
+     * internal hash map containing the {@link GroupedFlowables}.
```
`{@link GroupedFlowable}s`

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     *       CacheBuilder
+     *         .newBuilder() 
+     *         .maximumSize(3)
+     *         .removalListener(entry -> { 
```
`->` -> `-&gt;`

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * <p>
+     * <em>Note:</em> A {@link GroupedFlowable} will cache the items it is to emit until such time as it
+     * is subscribed to. For this reason, in order to avoid memory leaks, you should not simply ignore those
+     * {@code GroupedPublisher}s that do not concern you. Instead, you can signal to them that they may
```
`GroupedPublisher` -> `GroupedFlowable`

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * discard their buffers by applying an operator like {@link #ignoreElements} to them.
+     * <dl>
+     *  <dt><b>Backpressure:</b></dt>
+     *  <dd>Both the returned and its inner {@code Publisher}s honor backpressure and the source {@code Publisher}
```
`Publisher` -> `GroupedFlowable`

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     *  <dt><b>Backpressure:</b></dt>
+     *  <dd>Both the returned and its inner {@code Publisher}s honor backpressure and the source {@code Publisher}
+     *  is consumed in a bounded mode (i.e., requested a fixed amount upfront and replenished based on
+     *  downstream consumption). Note that both the returned and its inner {@code Publisher}s use
```
`Publisher` -> `GroupedFlowable`

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,112 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * @param evictingMapFactory
+     *            The factory used to create a map that will be used by the implementation to hold the 
+     *            {@link GroupedFlowable}s. The evicting map created by this factory must notify when a
+     *            value in this map has been evicted. The next source emission will bring about the 
```
How? By calling the provided input `Consumer<Object>` with something?

akarnokd commented on src/main/java/io/reactivex/internal/operators/flowable/FlowableGroupBy.java:
```diff
@@ -78,13 +95,21 @@ protected void subscribeActual(Subscriber<? super GroupedFlowable<K, V>> s) {
             this.delayError = delayError;
-            this.groups = new ConcurrentHashMap<Object, GroupedUnicast<K, V>>();
+            if (mapFactory == null) {
+                this.groups = new ConcurrentHashMap<Object, GroupedUnicast<K, V>>();
```
I'm not a fan of this structure. These could be constructor arguments prepared before the `GroupBySubscriber` gets initialized. So the try-catch will only cover the non-null case only and the use of `GroupBySubscriber` can look like before:

```java
source.subscribe(new GroupBySubscriber<T, K, V>(s, keySelector, valueSelector, bufferSize, delayError, groups, evictedGroups));
```

akarnokd commented on src/main/java/io/reactivex/internal/operators/flowable/FlowableGroupBy.java:
```diff
@@ -371,6 +408,27 @@ public boolean isEmpty() {
     }
+    
+    @SuppressWarnings("unchecked")
+    private static <K,V> Map<Object, GroupedUnicast<K, V>> createMap(
```
Not sure why this is a separate method. Also please avoid `private`.

davidmoten commented:
Thanks @akarnokd. I've addressed your comments and also added `@Beta` and `@since 2.1.11`.

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9787,9 +9787,10 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
-     *            completion of the evicted {@link GroupedFlowable}s. See example using Guava above.
+     *            {@link GroupedFlowable}s. The evicting map created by this factory must
+     *            notify the provided {@code Consumer<Object>} with the entry value (not the key!) when
+     *            an entry in this map has been evicted.. The next source emission will bring about the
```
Two dots.

akarnokd commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9798,10 +9799,13 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
      *         key value
      * @see <a href="http://reactivex.io/documentation/operators/groupby.html">ReactiveX operators documentation: GroupBy</a>
+     *
+     * @since 2.1.11
```
This should fit into 2.1.10, assuming @vanniktech or @artem-zinnatullin approves.

vanniktech commented on build.gradle:
```diff
@@ -43,6 +43,7 @@ apply plugin: "me.champeau.gradle.jmh"
 apply plugin: "com.github.hierynomus.license"
 apply plugin: "com.jfrog.bintray"
 apply plugin: "com.jfrog.artifactory"
+apply plugin: "eclipse"
```
why do we need the eclipse plugin?

davidmoten replied:
This is so I can run `./gradlew eclipse` for import into Eclipse without using gradle plugin. It's not required for the general build obviously but is a convenience for me. Can remove if has any negative effects.

vanniktech replied:
Works for me.

vanniktech commented on build.gradle:
```diff
@@ -55,7 +56,7 @@ def reactiveStreamsVersion = "1.0.2"
 def jmhLibVersion = "1.19"
 def testNgVersion = "6.11"
-
+def guavaVersion = "19.0"
```
Latest version is 24.0

davidmoten replied:
Yes I know but 19.0 will do. I don't plan to keep this test dependency up to date. No point as it is for demonstration purposes only.

vanniktech commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,116 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * 
+     * <p>The map created by an {@code evictingMapFactory} must be thread-safe.
+     * 
+     * <p>An example of an {@code evictingMapFactory} using <a href="https://google.github.io/guava/releases/19.0/api/docs/com/google/common/cache/CacheBuilder.html">CacheBuilder</a> from the Guava library is below:
```
Probably we should reference a newer version link

davidmoten replied:
Not that keen to track this against latest release of guava in the future but can go to v24 if required.

davidmoten replied:
have moved to 24.0

vanniktech commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,116 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * Flowable
+     *   .range(1, 1000)
+     *   // note that number of keys is 10
+     *   .groupBy(x -&gt; x % 10, x-&gt; x, true, 16, evictingMapFactory)
```
we probably want `x -&gt; x` here to be consistent with the first parameter

davidmoten replied:
don't understand that ??

vanniktech replied:
`x-&gt; x` becomes `x-> x`
`x -&gt; x` becomes `x -> x` which is then more readable and also consistent with the first parameter

vanniktech commented on src/main/java/io/reactivex/Flowable.java:
```diff
@@ -9708,9 +9708,116 @@ public final Disposable forEachWhile(final Predicate<? super T> onNext, final Co
+     * <p>
+     * <em>Note:</em> A {@link GroupedFlowable} will cache the items it is to emit until such time as it
+     * is subscribed to. For this reason, in order to avoid memory leaks, you should not simply ignore those
+     * {@code GroupedFlowable}s that do not concern you. Instead, you can signal to them that they may
```
`Instead, you can signal to them that they may` sounds very weird to me. Can we rephrase it somehow?

davidmoten replied:
this is a legacy comment that occurs in every `groupBy` overload. Perhaps we could leave rework of that phrase to another PR? I agree that the para is a bit clunky and could probably do with explicit  examples like don't filter on GroupedFlowable keys (`.filter(g -> g.getKey() > 10)`) do blah instead.

vanniktech replied:
Then leave as is.

vanniktech reviewed (approved):
Looking good. Let's get this into 2.1.10

davidmoten commented:
I've 

* updated the guava dependency to 24.0-jre
* updated the CacheBuilder link to point to Guava 24.0 javadoc
* set the since version to 2.1.10

davidmoten commented:
Added that missing space to javadoc (`x -> x`) and squashed commits. Thanks for the review @vanniktech.

vanniktech reviewed (approved):
:shipit:
